Add Google Vertex AI provider

rakibulrocky14 opened the pull request:
# Summary

* Adds first-class **Google Vertex AI provider** using `@ai-sdk/google-vertex`.
* Supports **Gemini 2.5** models and partner **MaaS (Model Garden)** models via full publisher IDs.
* New **Vertex-specific settings UI** for Project, Location, and Service Account JSON.
* Implements a **“thinking” toggle** for Gemini 2.5 Flash

  * Pro: always on
  * Flash: toggleable
  * Flash Lite: none
* Fixes *“AI not found”* for Vertex built-ins by mapping to `publishers/google` paths.
* Hardens **cross-platform file ops** and ensures all tests pass.

---

# What’s New

### Vertex AI Provider

* Uses `@ai-sdk/google-vertex` with `googleAuthOptions.credentials` from pasted Service Account JSON.
* Configurable **project** and **location**.
* Base URL → `/projects/{project}/locations/{location}`

  * Built-ins: `publishers/google/models/<id>`
  * Partner MaaS: `publishers/<partner>/models/...`

### Built-in Vertex Models

* `gemini-2.5-pro`
* `gemini-2.5-flash`
* `gemini-2.5-flash-lite`

### Thinking Behavior

* Vertex + Google marked as thinking-capable.
* Pro: always thinking
* Flash: toggle in UI
* Flash Lite: none

### Vertex Settings UI

* New **Google Vertex AI panel** for Project ID, Location, Service Account JSON.
* Keys encrypted like other secrets.

---

# Fixes

* **Model resolution:** built-ins auto-map to `publishers/google/models/<id>`.
* **Partner MaaS support:** full publisher IDs work directly (e.g. DeepSeek).
* **Cross-platform paths:** normalize file ops with `toPosixPath`, preserve `safeJoin` semantics.

---

# Why This Is Better

* Users can select **Vertex alongside other providers**.
* **More models** available through Model Garden.
* **Dedicated setup UI** reduces misconfig.
* **Thinking toggle** gives control over cost vs. reasoning depth.

---

# Files Changed

* **Provider & Models**: `language_model_helpers.ts`, `get_model_client.ts`
* **Streaming**: `chat_stream_handlers.ts`
* **Schemas & Encryption**: `schemas.ts`, `settings.ts`
* **Settings UI**: `VertexConfiguration.tsx`, `ApiKeyConfiguration.tsx`
* **Models UI**: `ModelsSection.tsx` (Flash toggle)
* **Setup Detection**: `useLanguageModelProviders.ts`
* **Path Utils**: `path_utils.ts`, `response_processor.ts`
* **Deps**: `package.json` → `@ai-sdk/google-vertex@3.0.16`

---

# Tests & Validation

* **TypeScript**: `npm run ts` → ✅
* **Lint**: `npm run lint` → ✅
* **Unit tests**: `npm test` → ✅ 231 passed, 0 failed

---

# Migration / Notes

* No breaking changes.
* For Vertex usage:

  * Ensure Vertex AI API is enabled.
  * Service Account needs `roles/aiplatform.user`.
  * Region must support model (e.g. `us-central1`).
* Thinking toggle currently affects **only** Gemini 2.5 Flash.

---

# Manual QA

1. Configure Vertex with Project/Location/Service Account JSON.
2. Test built-ins:

   * `gemini-2.5-pro`
   * `gemini-2.5-flash` (toggle on/off)
   * `gemini-2.5-flash-lite`
3. Test MaaS partner model (e.g., DeepSeek) via full publisher ID.
4. Verify other providers remain unaffected.
    
<!-- This is an auto-generated description by cubic. -->
---

## Summary by cubic
Adds a first-class Google Vertex AI provider with Gemini 2.5 models, a Vertex settings panel, and a “thinking” toggle for Gemini 2.5 Flash. Also fixes model resolution for Vertex and hardens cross-platform file operations.

- **New Features**
  - Vertex AI provider via @ai-sdk/google-vertex with project, location, and service account JSON.
  - Built-in models: gemini-2.5-pro, gemini-2.5-flash, gemini-2.5-flash-lite.
  - “Thinking” support: Pro always on; Flash toggle in Models UI; Flash Lite none.
  - MaaS partners supported via full publisher paths (e.g., publishers/<partner>/models/...).
  - Vertex settings UI with encrypted service account key storage.

- **Bug Fixes**
  - Built-in Vertex models auto-map to publishers/google/models/<id>.
  - Consistent file ops across platforms using toPosixPath.
  - Vertex readiness detection requires project/location/service account JSON.
  - Streaming “thinking” behavior respects Vertex Flash toggle and Pro always-on.

<!-- End of auto-generated description by cubic. -->

mdrirocky committed:
feat(vertex): add Vertex AI provider w/ Gemini 2.5, Flash thinking toggle, MaaS support, and path fixes

- Add Vertex provider via @ai-sdk/google-vertex
- Config: projectId, location, serviceAccountKey (encrypted)
- UI: Vertex settings panel + Flash “thinking” toggle
- Built-in models: gemini-2.5-pro | gemini-2.5-flash | gemini-2.5-flash-lite
- Support partner MaaS publisher models
- Add toPosixPath, fix Windows test paths
- Secure serviceAccountKey storage

mdrirocky committed:
Merge branch 'gc-vertex-imp' of https://github.com/rakibulrocky14/dyad into gc-vertex-imp

rakibulrocky14 committed:
Update src/ipc/processors/response_processor.ts

fs.rmdirSync() is deprecated in Node.js and scheduled for removal in future versions. So replacing with fs.rmSync() using { recursive: true, force: true } options for better cross-platform compatibility and future-proofing. This change would align with modern Node.js filesystem API recommendations.

Co-authored-by: graphite-app[bot] <96075541+graphite-app[bot]@users.noreply.github.com>

wwwillchen commented on src/ipc/utils/path_utils.ts:
```diff
@@ -58,3 +58,12 @@ export function safeJoin(basePath: string, ...paths: string[]): string {
   return joinedPath;
 }
+
+/**
```
we already have https://github.com/dyad-sh/dyad/blob/67c67e7bec76f56664a7ea7b48ccc829ca8a9dbd/shared/normalizePath.ts#L5

wwwillchen commented on src/lib/schemas.ts:
```diff
@@ -59,7 +60,17 @@ export type LargeLanguageModel = z.infer<typeof LargeLanguageModelSchema>;
 export const ProviderSettingSchema = z.object({
+  // Generic API key (used by most providers except Vertex/Azure)
   apiKey: SecretSchema.optional(),
+
```
this should be modeled as a union.
either a RegularProviderSettingSchema or a VertexProviderSettingSchema

wwwillchen commented on src/lib/schemas.ts:
```diff
@@ -59,7 +60,17 @@ export type LargeLanguageModel = z.infer<typeof LargeLanguageModelSchema>;
+  // Service account JSON key stored as a Secret (encrypted like apiKey)
+  serviceAccountKey: SecretSchema.optional(),
+  // Per-model toggle for Gemini 2.5 Flash thinking (Vertex only)
+  enableFlashThinking: z.boolean().optional(),
```
this doesn't seem right to me, we shouldn't put model-specific info in provider setting.

let's remove the flash thinking budget from this PR to keep this PR simpler. I think the thinking budget should (ideally) be handled in a model-agnostic, but it's quite tricky because not every model supports thinking budget, or even disabling thinking.

rakibulrocky14 replied:
@wwwillchen 
yeah I am not putting for all models for turn off and on, I added only for 2.5 flash for flexibility because you know that flash 2.5 thinking and non thinking have different pricing and maybe someone want to use non thinking for price that’s why just let’s keep the toggle

wwwillchen commented on src/ipc/processors/response_processor.ts:
i don't think we need to do path normalization here because we already do it in the tag parser: https://github.com/dyad-sh/dyad/blob/main/src/ipc/utils/dyad_tag_parser.ts

also, these changes don't seem related to vertex AI, and I'd prefer the PRs to be focused (and make it easier to review). thanks.

rakibulrocky14 replied:
@wwwillchen I actually didn’t want but on npm test it showed error so I fixed it, well, I will remove it

wwwillchen commented on src/ipc/handlers/chat_stream_handlers.ts:
```diff
@@ -667,6 +667,26 @@ This conversation includes one or more image attachments. When the user uploads
+          const selectedModelName = settings.selectedModel.name;
+          let includeGoogleThoughts: boolean | undefined = undefined;
+          if (selectedProviderId === "google") {
+            // Keep existing behavior for Google provider
```
see other comment - let's remove this thinking budget-related code from this PR - this doesn't seem like a very clean approach and will become a maintenance burden as there's new models.

rakibulrocky14 replied:
@wwwillchen
yeah I am not putting for all models for turn off and on, I added only for 2.5 flash for flexibility because you know that flash 2.5 thinking and non thinking have different pricing and maybe someone want to use non thinking for price that’s why just let’s keep the toggle

> Well, if you say then I will remove, but keeping this will be loved by the users believe me

wwwillchen replied:
@rakibulrocky14 during preview, flash2.5 thinking and non-thinking had different prices but now that it's in GA the prices are the same: https://cloud.google.com/vertex-ai/generative-ai/pricing#gemini-models-2.5 so my guess is most people would want to keep thinking on (since price is same and performance is much better with thinking)

rakibulrocky14 replied:
@wwwillchen Well, my bad, Thanks and I will change it

wwwillchen reviewed (changes requested):
thanks for the PR - left a few comments. I think overall PR makes sense but let's keep this focused on the core vertex AI provider support and leave alone thinking budget, path normalization for future PRs.

rakibulrocky14 commented:
@wwwillchen 

Well I will work on these and update the PR 

But 

> I think we should keep the flash thinking turn on and off toggle to provide users more flexibility

rakibulrocky14 commented:
@wwwillchen I will wait for your response

rakibulrocky14 committed:
PR #1163 review: remove Vertex Flash thinking toggle; keep Google thoughts always-on; refactor provider settings to union; revert unrelated path normalization; trim Vertex projectId and fix isConfigured

rakibulrocky14 committed:
Merge branch 'main' into gc-vertex-imp

rakibulrocky14 commented:
@wwwillchen check now

wwwillchen reviewed (commented):
I pulled it down but can't actually get it to work. the settings don't seem to be saving correctly, could you double check?

rakibulrocky14 committed:
Enhance VertexConfiguration and provider settings

- Added a helpful message in VertexConfiguration to guide users on region selection for partner models.
- Refactored provider options in chat_stream_handlers to streamline Google and OpenAI configurations, ensuring proper handling of model-specific settings.
- Updated schemas for provider settings to include passthrough for additional fields, improving flexibility in configuration.

This update improves user experience and maintains compatibility with various model configurations.

rakibulrocky14 commented:
@wwwillchen I hope now all will work as expected. I tested all models, including many MaaS Partner models like Qwen3 Coder, DeepSeek R1 0528, and they are all working well

rakibulrocky14 committed:
Enhance provider configuration handling in ProviderSettingsPage

- Added special handling for Vertex configuration status to determine if it is properly set up.
- Updated the isConfigured logic to account for Vertex alongside existing Azure and user key checks, improving clarity and functionality.

This change enhances the configuration validation process for different providers, ensuring a more robust settings management experience.

wwwillchen committed:
cleanups

wwwillchen reviewed (approved):
thanks! i pulled it down and it works now!
i made some changes to make it more type-safe (and also removed flash-lite from preset model since i can rarely get it output working code).

running CI and then will merge after

wwwillchen committed:
fix

wwwillchen committed:
clean
